source/common/stream_info/utility.cc

```diff
@@ -43,6 +49,231 @@ absl::optional<ResponseFlag> ResponseFlagUtils::toResponseFlag(absl::string_view
+    return 502;
+  case ProxyStatusError::DestinationIpUnroutable:
+    return 502;
+  case ProxyStatusError::ConnectionRefused:
+    return 502;
+  case ProxyStatusError::ConnectionTerminated:
+    return 502;
+  case ProxyStatusError::ConnectionTimeout:
+    return 504;
+  case ProxyStatusError::ConnectionReadTimeout:
+    return 504;
+  case ProxyStatusError::ConnectionWriteTimeout:
+    return 504;
+  case ProxyStatusError::ConnectionLimitReached:
+    return 503;
+  case ProxyStatusError::TlsProtocolError:
+    return 502;
+  case ProxyStatusError::TlsCertificateError:
+    return 502;
+  case ProxyStatusError::TlsAlertReceived:
+    return 502;
+  case ProxyStatusError::HttpRequestDenied:
+    return 403;
+  case ProxyStatusError::HttpResponseIncomplete:
+    return 502;
+  case ProxyStatusError::HttpResponseHeaderSectionSize:
+    return 502;
+  case ProxyStatusError::HttpResponseHeaderSize:
+    return 502;
+  case ProxyStatusError::HttpResponseBodySize:
+    return 502;
+  case ProxyStatusError::HttpResponseTrailerSectionSize:
+    return 502;
+  case ProxyStatusError::HttpResponseTrailerSize:
+    return 502;
+  case ProxyStatusError::HttpResponseTransferCoding:
+    return 502;
+  case ProxyStatusError::HttpResponseContentCoding:
+    return 502;
+  case ProxyStatusError::HttpResponseTimeout:
+    return 504;
+  case ProxyStatusError::HttpUpgradeFailed:
+    return 502;
+  case ProxyStatusError::HttpProtocolError:
+    return 502;
+  case ProxyStatusError::ProxyInternalError:
+    return 500;
+  case ProxyStatusError::ProxyConfigurationError:
+    return 500;
+  case ProxyStatusError::ProxyLoopDetected:
+    return 502;
+  case ProxyStatusError::ProxyInternalResponse:
+  case ProxyStatusError::HttpRequestError:
+  default:
+    return absl::nullopt;
+  }
+}
+
+const std::string
+ProxyStatusUtils::toString(const StreamInfo& stream_info, const ProxyStatusError error,
+                           absl::string_view server_name,
+                           const envoy::extensions::filters::network::http_connection_manager::v3::
+                               HttpConnectionManager::ProxyStatusConfig& proxy_status_config) {
+  std::string retval = "";
+
+  switch (proxy_status_config.proxy_name()) {
+  case envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager::
+      ProxyStatusConfig::SERVER_NAME: {
+    retval.append(std::string(server_name));
+    break;
+  }
+  case envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager::
+      ProxyStatusConfig::ENVOY_LITERAL:
+  default: {
+    retval.append(Envoy::Http::DefaultServerString::get());
+    break;
+  }
+  }
+
+  retval.append(absl::StrFormat("; error=%s", proxyStatusErrorToString(error)));
+
+  if (!proxy_status_config.remove_details() && stream_info.responseCodeDetails().has_value()) {
+    retval.append(absl::StrFormat("; details='%s'", stream_info.responseCodeDetails().value()));
```

Comment:
I think it would be good to include `connectionTerminationDetails()` as well. We could also list the original flags here as well, since we loose some precision when it is converted to standard error.

Reply:
Good idea -- I've added these details to the details="" field (and added a new ProxyStatusConfig bool to disable them). I'll do the same for the original flags, and then resolve this conversation.

Reply:
Done.